fuse_models/include/fuse_models/unicycle_2d_state_cost_functor.h

```diff
@@ -147,38 +143,50 @@ bool Unicycle2DStateCostFunctor::operator()(
   const T* const acc_linear2,
   T* residual) const
 {
-  T position_pred[2];
-  T yaw_pred[1];
+  T delta_position_pred[2];
+  T delta_yaw_pred[1];
   T vel_linear_pred[2];
   T vel_yaw_pred[1];
   T acc_linear_pred[2];
+  T init_position[2];
+  T init_yaw[1];
+  init_position[0] = T(0.0);
+  init_position[1] = T(0.0);
+  init_yaw[0] = T(0.0);
   predict(
-    position1,
-    yaw1,
+    init_position,
+    init_yaw,
     vel_linear1,
     vel_yaw1,
     acc_linear1,
     T(dt_),
-    position_pred,
-    yaw_pred,
+    delta_position_pred,
+    delta_yaw_pred,
     vel_linear_pred,
     vel_yaw_pred,
     acc_linear_pred);
 
-  Eigen::Map<Eigen::Matrix<T, 8, 1>> residuals_map(residual);
-  residuals_map(0) = position2[0] - position_pred[0];
-  residuals_map(1) = position2[1] - position_pred[1];
-  residuals_map(2) = yaw2[0] - yaw_pred[0];
-  residuals_map(3) = vel_linear2[0] - vel_linear_pred[0];
-  residuals_map(4) = vel_linear2[1] - vel_linear_pred[1];
-  residuals_map(5) = vel_yaw2[0] - vel_yaw_pred[0];
-  residuals_map(6) = acc_linear2[0] - acc_linear_pred[0];
-  residuals_map(7) = acc_linear2[1] - acc_linear_pred[1];
+  const T delta_yaw_est = yaw2[0] - yaw1[0];
+  const Eigen::Matrix<T, 2, 1> position1_map(position1[0], position1[1]);
+  const Eigen::Matrix<T, 2, 1> position2_map(position2[0], position2[1]);
+  const Eigen::Matrix<T, 2, 1> position_diff = (position2_map - position1_map);
+  const T sin_yaw_inv = ceres::sin(-yaw1[0]);
+  const T cos_yaw_inv = ceres::cos(-yaw1[0]);
+
+  residual[0] = (cos_yaw_inv * position_diff.x() - sin_yaw_inv * position_diff.y()) - delta_position_pred[0];
+  residual[1] = (sin_yaw_inv * position_diff.x() + cos_yaw_inv * position_diff.y()) - delta_position_pred[1];
```

Comment:
Same comment as before, although this suggestion might not work as is when `T` is a `ceres::Jet`, so it probably needs some changes:
```suggestion
  const Eigen::Matrix<T, 2, 1> delta_position_delta(delta_position_delta_x, delta_position_delta_y);
  const Eigen::Matrix<T, 2, 2> R_yaw_inv = fuse_core::rotationMatrix2D(-parameters[1][0]);

  residuals.head<2>() = R_yaw_inv * position_diff - delta_position_delta;
```

If it's too hard to make it work, just leave it as is. For the 2D rotation matrix I believe we'd need something like this: https://github.com/ceres-solver/ceres-solver/blob/0a7845fef4a87bc803a3e25e198251bf55858bfd/examples/slam/pose_graph_2d/pose_graph_2d_error_term.h#L42-L50 Maybe `fuse_core` already have that or otherwise should have it.